utoipa-gen/src/path/response.rs

```diff
@@ -138,26 +158,33 @@ impl ToTokens for Response<'_> {
     }
 }
 
-impl ContentTypeResolver for Response<'_> {}
+impl ContentTypeResolver for ResponseValue<'_> {}
 
 pub struct Responses<'a>(pub &'a [Response<'a>]);
 
 impl ToTokens for Responses<'_> {
     fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
-        if self.0.is_empty() {
-            tokens.extend(quote! { utoipa::openapi::Responses::new() })
-        } else {
-            let responses = self.0.iter().fold(quote! {}, |mut acc, response| {
-                let code = &response.status_code.to_string();
-                acc.extend(quote! { (#code, #response), });
+        tokens.extend(self.0.iter().fold(
+            quote! { utoipa::openapi::ResponsesBuilder::new() },
+            |mut acc, response| {
+                match response {
+                    Response::IntoResponses(path) => {
+                        let span = path.span();
+                        acc.extend(quote_spanned! {span =>
+                            .responses_from_into_responses::<#path>()
+                        })
+                    }
+                    Response::Value(response) => {
+                        let code = &response.status_code.to_string();
+                        acc.extend(quote! { .response(#code, #response) });
+                    }
+                }
 
                 acc
-            });
+            },
+        ));
 
-            tokens.extend(quote! {
-                utoipa::openapi::Responses::from_iter([#responses])
-            });
-        }
```

Comment:
I guess we cannot use similar approach here since there might be `into_responses()` call for responses and same time a `Response` object. The nice thing with this `from_iter()` is that it would make only one allocation of exact size. 

When looping over items without creating a map or a vec with extact size with `::with_capacity()` would make the result collection to reallocate when items where added.

Reply:
Thanks I didn't think of that. [It looks like there is no `::with_capacity()` for `BTreeMap`](https://users.rust-lang.org/t/btreemap-with-capacity/39770/3) which is [used by us to store responses](https://github.com/juhaku/utoipa/pull/215/files#diff-eea3ccce264a9afded89ef39496101cb1a5bf8c1fedc4cd4ddf9221e4744d922R27).

It also doesn't make use of the size hint/length anywhere with its `FromIterator` that I can find:
+ [BTreeMap FromIterator impl](https://stdrs.dev/nightly/x86_64-unknown-linux-gnu/src/alloc/collections/btree/map.rs.html#2089-2099)
+ [BTreeMap internal bulk_push](https://stdrs.dev/nightly/x86_64-unknown-linux-gnu/src/alloc/collections/btree/append.rs.html#38-90) (looks like it doesn't use the size hint anywhere)

We could switch to using a Vec<(String, Response)> internally for `Responses` but we would lose the deduplication guarantees. What do you think?

Reply:
Anyhow I can add a `From<IntoResponses>` impl for `Responses` for people who want it.

Reply:
The other thing about this is that currently this macro supports interleaving both `IntoResponses` and regular tuple defined responses (just like `IntoParams` for `params()`), which is why `responses.extend()` is used in the impl for  `ResponsesBuilder::responses_from_into_responses()`:

```rust
responses(
    Response,
    (status = 404, description = "some extra response"),
)
```

Reply:
Another thing I'm wondering is that our status codes are currently strings, but should we perhaps enforce them as `u16`? Or is the http status code fair game for any string in practice? It doesn't seem like the openapi spec itself enforces this.

Reply:
> We could switch to using a Vec<(String, Response)> internally for Responses but we would lose the deduplication guarantees. Or we could switch to using a HashMap<String, Response> but we would lose the ordering. These types make use of the size hint in their FromIterator to avoid allocations as I understand it. What do you think?

I think we can keep it as is with the `BTreeMap`. I don't think the benefits would out weight the ordering support and be that significant. So I guess we can keep it as is? Any thoughts?

Reply:
> We would have to first collect all of these together in a single data structure, and then use FromIterator for Responses. Even then I'm not sure if we could compose it down to a single allocation.

Yeah might just be a waste of effort to try to reduce allocations in this case

Reply:
> I think we can keep it as is with the BTreeMap. I don't think the benefits would out weight the ordering support and be that significant. So I guess we can keep it as is? Any thoughts?

I think it sounds reasonable to keep as is, we can find a way to optimise it later if we need.

> We could internally use it as u16 might actually not be a bad idea. I originally thought to use a number type for this. But leaned back to String because within OpenAPI document the responses must string=response map.

Nice I'll make a new issue for this idea :slightly_smiling_face:

Reply:
Sounds reasonable.

Reply:
> Nice I'll make a new issue for this idea slightly_smiling_face, and perhaps implement it in this PR because otherwise we'll have to break the interface again.

I think I'll abandon this idea, it requires a custom serializer/deserialzer implementation, can implement it in a future PR